c++/src/capnp/rpc.c++

```diff
@@ -3179,16 +3182,10 @@ private:
   }
 
   kj::Promise<void> acceptLoop() {
-    auto receive = network.baseAccept().then(
+    return network.baseAccept().then(
         [this](kj::Own<VatNetworkBase::Connection>&& connection) {
       getConnectionState(kj::mv(connection));
-    });
-    return receive.then([this]() {
-      // No exceptions; continue loop.
-      //
-      // (We do this in a separate continuation to handle the case where exceptions are
-      // disabled.)
-      tasks.add(acceptLoop());
```

Comment:
It's entirely possible that I'm missing something about this continuation, I don't know much about exceptions being disabled.

Also possible that I'm missing something about the previous use of `tasks` here that differs from storing a promise loop?

Reply:
The `tasks.add(acceptLoop())`thing is because this code actually predates promise tail call optimization, so I had to do it that way to avoid a memory leak. `return acceptLoop()` is strictly better and we should have changed that a long time ago.

As for the exceptions issue: If `getConnectionState()` throws a recoverable exception, but the library is compiled with `-fno-exceptions`, then `getConnectionState()` will still return normally. Once the contituation completes, the event loop will notice the exception was raised, and will discard the continuation result in favor of the exception. But if the continuation calls `tasks.add(acceptLoop())` immediately after `getConnectionState()`, this will be executed even when `getConnectionState()` threw, and hence the loop would continue despite the exception. However, now that you've changed it to `return acceptLoop()`, I think that problem is solved. (Technically `acceptLoop()` will be invoked an extra time and then immediately canceled, but that's not the kind of thing I worry too much about when it comes to supporting no-exceptions.)

So I think your changes here are fine.

Reply:
Nice, thanks for the explanation!